File: python/src/tradingflow/operators/filter.py
"""Filter operator -- predicate-gated passthrough."""

from __future__ import annotations

from collections.abc import Callable
from dataclasses import dataclass
from typing import Any

import numpy as np

from ..operator import Operator, Notify
from ..types import Array, Handle, NodeKind


@dataclass(slots=True)
class FilterState[T: np.generic]:
    """State for [`Filter`]."""

    predicate: Callable[[np.ndarray], bool]


class Filter[T: np.generic](
    Operator[
        tuple[Handle[Array[T]]],
        Handle[Array[T]],
        FilterState[T],
    ]
):
    """Passes or drops the entire element based on a predicate.

    When the predicate returns `False`, the operator produces no output
    for that timestamp, halting downstream propagation.

    Parameters
    ----------
    input
        Upstream handle.
    predicate
        Callable receiving a numpy array (the current value) and returning
        `True` to pass it through, `False` to drop.
    name
        Optional human-readable name.
    """

    __slots__ = ("_predicate",)

    def __init__(
        self,
        input: Handle[Array[T]],
        predicate: Callable[[np.ndarray], bool],
        *,
        name: str | None = None,
    ) -> None:
        self._predicate = predicate
        super().__init__(
            inputs=(input,),
            kind=NodeKind.ARRAY,
            dtype=input.dtype,
            shape=input.shape,
            name=name,
        )

    def init(self, inputs: tuple, timestamp: int) -> FilterState[T]:
        return FilterState(predicate=self._predicate)

    @staticmethod
    def compute(
        state: FilterState,
        inputs: tuple,
        output: Any,
        timestamp: int,
        notify: Notify,
    ) -> bool:
        value = inputs[0].value()
        if state.predicate(value):
            output.write(value)
            return True
        return False

File: src/experimental_alt/operators/filter.rs
//! Predicate-gated filter — ported from `src/operators/filter.rs`.

use std::marker::PhantomData;

use super::super::data::{Array, Input, InputTypes, Instant, Scalar};
use super::super::operator::Operator;

pub struct Filter<T: Scalar, F: Fn(&Array<T>) -> bool> {
    predicate: F,
    _phantom: PhantomData<T>,
}

impl<T: Scalar, F: Fn(&Array<T>) -> bool> Filter<T, F> {
    pub fn new(predicate: F) -> Self {
        Self {
            predicate,
            _phantom: PhantomData,
        }
    }
}

impl<T: Scalar, F: Fn(&Array<T>) -> bool + Send + 'static> Operator for Filter<T, F> {
    type State = Self;
    type Inputs = Input<Array<T>>;
    type Output = Array<T>;

    fn init(self, inputs: &Array<T>, _timestamp: Instant) -> (Self, Array<T>) {
        (self, inputs.clone())
    }

    #[inline(always)]
    fn compute(
        state: &mut Self,
        inputs: &Array<T>,
        output: &mut Array<T>,
        _timestamp: Instant,
        _produced: <Self::Inputs as InputTypes>::Produced<'_>,
    ) -> bool {
        if (state.predicate)(inputs) {
            output.as_mut_slice().clone_from_slice(inputs.as_slice());
            true
        } else {
            false
        }
    }
}
